Tests/SystemTests/Utils/AuthHelper.swift

```diff
@@ -0,0 +1,36 @@
+//
+//  File.swift
+//  
+//
+//  Created by Łukasz Szyszkowski on 16/08/2021.
+//
+
+import Foundation
+import AblyAssetTrackingCore
+import Ably
+import XCTest
+
+class AuthHelper {
+    func requestToken(options: ARTClientOptions, clientId: String? = nil) -> TokenDetails? {
+        var requestCompleted = false
+        var fetchedTokenDetails: TokenDetails?
+        let client = ARTRest(options: options)
+        let tokenParams = ARTTokenParams(clientId: clientId)
+        
+        client.auth.requestToken(tokenParams, with: nil) { tokenDetails, error in
+            if let error = error {
+                XCTFail("TokenDetails request failed with error: \(error)")
+            } else {
+                fetchedTokenDetails = tokenDetails?.asTokenDetails()
+            }
+          
+            requestCompleted = true
+        }
+        
+        while !requestCompleted {
+            CFRunLoopRunInMode(CFRunLoopMode.defaultMode, CFTimeInterval(0.1), false)
+        }
+        
+        return fetchedTokenDetails
+    }
+}
```

Comment:
This is an interesting helper method. Would this be a good idea for applications (non test code?)

Reply:
What did you find interesting here? :)

Reply:
Mostly `CFRunLoopRunInMode`, I was wondering when you should use this? And if there are other options here, e.g. semaphore's, which are more conventionally used. https://stackoverflow.com/a/40491615/7365866

Reply:
This code is copy-paste from `ably-cocoa` SDK.

We can't use a semaphore here, because the intention is to block the main thread https://stackoverflow.com/a/54104259
I guess `CFRunLoopRunInMode` config is just for slowing down run loop

Reply:
> This code is copy-paste from ably-cocoa SDK.

Okay 😂. I'm not sure what you mean by **slowing down run loop**. I'm trying to understand that line but I don't have any understanding at the moment.

Do you know what `CFRunLoopRunInMode` **does**? It looks like it's used to block the main thread, just like a semaphore would. I think a semaphore would be more obvious